### DISCUSSION TITLE
"Message queue listen OAuth token" while replacing self-hosted runner

### DISCUSSION THREAD

**Discussion by AllDmeat** [community/community#64812] [Actions]
### Select Topic Area

Bug

### Body

Few days ago we noticed 5 out of 6 of our self-hosted runners started throwing exception during registration.

 After further investigation we've discovered, that it's somehow related to replacing existing offline runners:
- If I remove existing offline runner manually new one successfully takes its place.
- If I can't remove existing offline runner due to some problem — runner can't replace it either. And when I'm saying 'some problem' I really do mean that:
<img width="379" alt="Screenshot 2023-08-24 at 10 44 20" src="https://github.com/community/community/assets/5537636/3a3dd550-0d15-4dd4-aee4-72df9b75f6df">

I'm registering runner like this:
`./config.sh --url https://github.com/*** --ephemeral --replace --labels $ALL_LABELS --name $RUNNER_NAME --runnergroup macos --work _work --token $RUNNER_TOKEN`

## Runner Version and Platform
macOS 13.2.1 / macOS 13.5.1
Apple Silicon
Runner 2.308.0, arm64

## What's not working?
`Message queue listen OAuth token`
<img width="356" alt="Screenshot 2023-08-24 at 11 10 32" src="https://github.com/community/community/assets/5537636/adc1f55d-ac0a-484a-a372-fd5d506bc03a">

## Runner and Worker's Diagnostic Logs
[Success log, when I manually remove runner from organization before registering once again](https://github.com/community/community/files/12425923/success.log)
[Failure log, when runner tries to replace itself from previous run](https://github.com/community/community/files/12425922/failure.log)

**Comment by AllDmeat** [comment]
Figured out steps to reproduce:
1. Start the runner in your shell. Doesn't matter if it's running the job or idle.
2. Kill the shell with the runner.
3. Try to restart the runner.

I don't understand why a runner can't replace his previous self in such a case.

**Comment by OrfeasZ** [comment]
I'm also getting this issue since a couple of days ago on the Windows runner.

A workaround that has helped in my setup (runners start in a container, config themselves in a way similar to yours, serve a job, and then exit and get re-created anew):

1. Remove the `--ephemeral` flag.
2. Set the `ACTIONS_RUNNER_HOOK_JOB_COMPLETED` environment variable to a command that kills the runner.

Keep in mind that if you're not running this in an ephemeral filesystem like that of a container's, you might need to delete some of the dotfiles (e.g. `.credentials` and `.runner`) in the runner directory before being able to run `config` again, probably as part of the job completed step.

**Comment by james-rant** [comment]
I've also encountered this exact issue this morning.
We run VMs on a Mac host, and when I need to edit the VM base, I stop the running VMs, edit, then restart, and in the past this hasn't been a problem -- the VM boots, and the runner replaces the old one and continues as normal. But today, I did the same process as I've done before but this time upon restarting the VMs, they were unable to replace the old runners, with the same error `Message queue listen OAuth token`.

I had to go into the organisation settings and manually remove the runners before the VMs were able to create them again.

Edit to add:
It seems if the Github Action completes normally, the script cleanly removes the runner before restarting the VM and registering a new runner, so the replacement works in this case -- it's when the runner is stopped unexpectedly that this issue seems to occur.

**Comment by ghost** [comment]
Similar behavior this morning. Deleting manually each runner from organization worked.

**Comment by AvivPeledTalon** [comment]
I've been dealing with this for the past couple of days.
I have a similar setup with VMs which I register as Self Hosted Github Runners. Host got turned off unexpectedly and the machines wouldn't come back up because of this.
Manually removing the offline runners did the trick.

Github please do something about this!

**Comment by AllDmeat** [comment]
Resolved for me today. Checked it on different machines for multiple times — after a few retries runner successfully replaces the runner and starts listening for the jobs.

![Screenshot 2023-08-31 at 21 21 29](https://github.com/community/community/assets/5537636/8e863636-8660-4253-976a-fbfb70cadf87)

**Reply by james-rant** [reply]
Same here, thanks for letting us know!
